tests/Unit/DataStructures/Tensor/Expressions/Test_Contract.cpp

```diff
@@ -0,0 +1,744 @@
+                    SpatialIndex<3, UpLo::Lo, Frame::Inertial>>>
+      Slull(2_st);
+  assign_unique_datavector_tensor_values(Slull);
+
+  const Tensor<DataVector, Symmetry<2, 1>,
+               index_list<SpatialIndex<3, UpLo::Lo, Frame::Inertial>,
+                          SpatialIndex<4, UpLo::Lo, Frame::Inertial>>>
+      SkJij_contracted = TensorExpressions::evaluate<ti_k_t, ti_i_t>(
+          Slull(ti_k, ti_J, ti_i, ti_j));
+
+  for (size_t k = 0; k < 3; k++) {
+    for (size_t i = 0; i < 4; i++) {
+      DataVector expected_sum(2, 0.0);
+      for (size_t j = 0; j < 3; j++) {
+        expected_sum += Slull.get(k, j, i, j);
+      }
+      CHECK(SkJij_contracted.get(k, i) == expected_sum);
+    }
+  }
+
+  // Contract third and fourth indices of <3, 2, 2, 1> symmetry rank 4 (upper,
+  // lower, lower, upper) tensor to rank 2 tensor
+  Tensor<double, Symmetry<3, 2, 2, 1>,
+         index_list<SpacetimeIndex<4, UpLo::Up, Frame::Inertial>,
+                    SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>,
+                    SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>,
+                    SpacetimeIndex<3, UpLo::Up, Frame::Inertial>>>
+      Rullu{};
+  std::iota(Rullu.begin(), Rullu.end(), 0.0);
+
+  const Tensor<double, Symmetry<2, 1>,
+               index_list<SpacetimeIndex<4, UpLo::Up, Frame::Inertial>,
+                          SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>>>
+      RFcgG_contracted = TensorExpressions::evaluate<ti_F_t, ti_c_t>(
+          Rullu(ti_F, ti_c, ti_g, ti_G));
+
+  for (size_t f = 0; f < 5; f++) {
+    for (size_t c = 0; c < 4; c++) {
+      double expected_sum = 0.0;
+      for (size_t g = 0; g < 4; g++) {
+        expected_sum += Rullu.get(f, c, g, g);
+      }
+      CHECK(RFcgG_contracted.get(f, c) == expected_sum);
+    }
+  }
+
+  Tensor<DataVector, Symmetry<3, 2, 2, 1>,
+         index_list<SpacetimeIndex<4, UpLo::Up, Frame::Inertial>,
+                    SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>,
+                    SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>,
+                    SpacetimeIndex<3, UpLo::Up, Frame::Inertial>>>
+      Sullu(1_st);
+  assign_unique_datavector_tensor_values(Sullu);
+
+  const Tensor<DataVector, Symmetry<2, 1>,
+               index_list<SpacetimeIndex<4, UpLo::Up, Frame::Inertial>,
+                          SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>>>
+      SFcgG_contracted = TensorExpressions::evaluate<ti_F_t, ti_c_t>(
+          Sullu(ti_F, ti_c, ti_g, ti_G));
+
+  for (size_t f = 0; f < 5; f++) {
+    for (size_t c = 0; c < 4; c++) {
+      DataVector expected_sum(1, 0.0);
+      for (size_t g = 0; g < 4; g++) {
+        expected_sum += Sullu.get(f, c, g, g);
+      }
+      CHECK(SFcgG_contracted.get(f, c) == expected_sum);
+    }
+  }
+
+  // Contract first and fourth indices of <2, 1, 1, 1> symmetry rank 4 (upper,
+  // lower, lower, lower) tensor to rank 2 tensor and reorder indices
+  Tensor<double, Symmetry<2, 1, 1, 1>,
```

Comment:
This test case and the test case below test contracting and then swapping the order of the non-contracted indices. I only did this for this case (1st and 4th indices contracted, order of 2nd and 3rd swapped) and the below case (2nd and 3rd indices contracted, order of 1st and 4th indices swapped). Let me know if this should also be done for other cases like the following, or if this is sufficient:
- 1st and 2nd contracted, 3rd and 4th swapped
- 1st and 3rd contracted, 2nd and 4th swapped
- 2nd and 3rd contracted, 1st and 4th swapped
- 3rd and 4th contracted, 1st and 2nd swapped

Reply:
I've now gone ahead and added these in, anyway, in 2accf4e